justfile

```diff
@@ -1,55 +1,100 @@
+# just has no idiom for setting an default value for an environment variable
+# so we shell out, as we need VIRTUAL_ENV in the justfile environment
+export VIRTUAL_ENV  := `echo ${VIRTUAL_ENV:-.venv}`
+
+# TODO: make it /scripts on windows?
+export BIN := VIRTUAL_ENV + "/bin"
+export PIP := BIN + "/python -m pip"
+# enforce our chosed pip compile flags
+export COMPILE := BIN + "/pip-compile --allow-unsafe --generate-hashes"
+
+
 # list available commands
 default:
-    @just --list
+    @{{ just_executable() }} --list
+
+
+# clean up temporary files
+clean:
+    rm -rf .venv
 
-# Set up local dev environment
-dev_setup:
+
+# ensure valid virtualenv
+virtualenv:
     #!/usr/bin/env bash
-    set -euo pipefail
-    . scripts/setup_functions
-    dev_setup
+    # allow users to specify python version in .env
+    PYTHON_VERSION=${PYTHON_VERSION:-python3.9}
+    # create venv and upgrade pip
+    test -d $VIRTUAL_ENV || { $PYTHON_VERSION -m venv $VIRTUAL_ENV && $PIP install --upgrade pip; }
+    # ensure we have pip-tools so we can run pip-compile
+    test -e $BIN/pip-compile || $PIP install pip-tools
+
+
+# update requirements.prod.txt if requirement.prod.in has changed
+requirements-prod: virtualenv
+    @test requirements.prod.in -ot requirements.prod.txt || $COMPILE --output-file=requirements.prod.txt requirements.prod.in
+
+
+# update requirements.dev.txt if requirements.dev.in has changed
+requirements-dev: virtualenv
+    @test requirements.dev.in -ot requirements.dev.txt || $COMPILE --output-file=requirements.dev.txt requirements.dev.in
 
-# run the test suite. Optional args are passed to pytest
-test ARGS="":
+
+# ensure prod requirements installed and up to date
+prodenv: requirements-prod
     #!/usr/bin/env bash
-    set -euo pipefail
-    . scripts/setup_functions
-    dev_setup
+    # have the requirements changed since we last installed them?
+    test requirements.prod.txt -nt $VIRTUAL_ENV/.prod || exit 0
+    $PIP install -r requirements.prod.txt
+    touch $VIRTUAL_ENV/.prod
+    # force dev rebuild too
+    touch requirements.dev.in
 
-    python -m pytest --cov=. --cov-report html --cov-report term-missing:skip-covered {{ ARGS }}
 
-# runs the format (black), sort (isort) and lint (flake8) check but does not change any files
-check:
+# && dependencies are run after the recipe has run. Needs just>=0.9.9
+#
+# ensure dev requirements installed and up to date
+devenv: prodenv requirements-dev && install-precommit
     #!/usr/bin/env bash
-    set -euo pipefail
-    . scripts/setup_functions
-    dev_setup
+    # have the requirements changed since we last installed them?
+    test requirements.dev.txt -nt $VIRTUAL_ENV/.dev || exit 0
+    $PIP install -r requirements.dev.txt
+    touch $VIRTUAL_ENV/.dev
```

Comment:
What does this line do?

Reply:
it updates the $VIRTUAL_ENV/.dev file's mtime to now. That way its newer than the `requirements.dev.txt`, so it won't reinstall again until the .txt changes.

Reply:
Oh you're testing it a couple of lines above, I completely missed that 👍

Reply:
right. In make, this is a common and useful pattern, to use a specific hidden file as a timestamp for when the outputs don't have a nice single file you can otherwise use.

It'd be spelt similarly in make: 

```
$(VIRTUAL_ENV)/.dev: requirements.dev.txt
    $(PIP) install -r $<
    touch $@
```

But we have to do the file mtime comparison ourselves, as just doesn't do it.